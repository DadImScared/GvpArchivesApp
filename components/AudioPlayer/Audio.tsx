import * as React from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";

import { PlaybackStatus } from "expo";

import { audioPlayer } from "../../actions";
import { IReducerState } from "../../reducers";
import { AudioPlayerStateAndActions, getAudioPlayerData } from "../../reducers/audioPlayer";
import AudioPlayer from "../NativeAudioPlayer";

interface IProps extends AudioPlayerStateAndActions<"playing" | "seekTo"> {
  uri: string;
  onPlaybackStatusUpdate: (status: PlaybackStatus) => void;
}

export class Audio extends React.Component<IProps> {

  onPlaybackUpdate = (status: PlaybackStatus) => {
    // prevents duration slider from flashing when pressed
    if (this.props.playing) {
      this.props.onPlaybackStatusUpdate(status);
    }
  };

  render() {
    const { playing, uri, seekTo } = this.props;
    return (
      <AudioPlayer
        onPlaybackStatusUpdate={this.onPlaybackUpdate}
        playing={playing}
        seekTo={seekTo}
        uri={uri}
      />
    );
  }
}

const mapStateToProps = (state: IReducerState) => getAudioPlayerData(state);

const mapDispatchToProps = (dispatch: Dispatch) => ({
  onPlaybackStatusUpdate: (status: PlaybackStatus) => dispatch(audioPlayer.updatePlaybackInstance(status))
});

export default connect(mapStateToProps, mapDispatchToProps)(Audio) as React.ComponentType;
