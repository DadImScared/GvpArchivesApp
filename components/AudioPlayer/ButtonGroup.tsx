import * as React from "react";
import { View } from "react-native";
import { connect } from "react-redux";
import { Dispatch } from "redux";

import { Button } from "native-base";

import { audioPlayer } from "../../actions";
import { IReducerState } from "../../reducers";
import { AudioPlayerStateAndActions, getButtonGroupData } from "../../reducers/audioPlayer";
import Icon from "../Icon";

interface IButtonGroupProps extends AudioPlayerStateAndActions<"playing" | "togglePlaying"> {}

export class ButtonGroup extends React.Component<IButtonGroupProps> {
  render() {
    const { playing, togglePlaying } = this.props;
    return (
      <View style={{ justifyContent: "center", flexDirection: "row" }}>
        {/*
            native base button sets backgroundColor to null but this crashes on android
            fix by setting backgroundColor to empty string
         */}
        <Button transparent={true} style={{ backgroundColor: "" }} primary={true} onPress={togglePlaying}>
          <Icon style={{ color: "#3F51B5"}} name={playing ? "pause" : "play"} />
        </Button>
      </View>
    );
  }
}

const mapStateToProps = (state: IReducerState) => getButtonGroupData(state);

const mapDispatchToProps = (dispatch: Dispatch) => ({
  togglePlaying: () => dispatch(audioPlayer.togglePlaying())
});

export default connect(mapStateToProps, mapDispatchToProps)(ButtonGroup) as React.ComponentType;
