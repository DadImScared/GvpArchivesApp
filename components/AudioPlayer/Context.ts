import * as React from "react";

export const AudioPlayerContext = React.createContext({});
