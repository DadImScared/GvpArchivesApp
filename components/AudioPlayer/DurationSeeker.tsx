import * as React from "react";
import { Slider, Text, View } from "react-native";

interface IDurationSeekerProps {
  isSeeking: boolean;
  getSeekSliderPosition: () => number;
  getTimeStamp: () => string;
  onSeekSliderChange: () => void;
  onSeekSliderComplete: (value: number) => void;
}

export const DurationSeeker: React.FunctionComponent<IDurationSeekerProps> = (
  {
    getTimeStamp,
    getSeekSliderPosition,
    onSeekSliderChange,
    onSeekSliderComplete,
  }
) => (
  <View style={{ flexDirection: "row", justifyContent: "center", alignItems: "center", paddingTop: 12 }}>
    <Slider
      value={getSeekSliderPosition()}
      onValueChange={onSeekSliderChange}
      onSlidingComplete={onSeekSliderComplete}
      style={{ width: "65%" }}
    />
    <Text>{getTimeStamp()}</Text>
  </View>
);

export default DurationSeeker;
