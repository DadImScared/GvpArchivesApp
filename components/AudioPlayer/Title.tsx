import * as React from "react";
import { View } from "react-native";
import { connect } from "react-redux";

import Ticker from "../../components/Marquee";
import { IReducerState } from "../../reducers";
import { getTitleData } from "../../reducers/audioPlayer";

interface ITitleProps {
  songName: string;
}

export const Title: React.FunctionComponent<ITitleProps> = ({ songName }) => {
  return (
    <View style={{ justifyContent: "center", alignItems: "center" }}>
      <Ticker duration={5000} bounce={false} loop={true}>
        {songName}
      </Ticker>
    </View>
  );
};

const mapStateToProps = (state: IReducerState) => getTitleData(state);

export default connect(mapStateToProps)(Title) as React.ComponentType;
