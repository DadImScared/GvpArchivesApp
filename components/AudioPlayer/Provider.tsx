
import * as React from "react";

import { Audio, PlaybackObject, PlaybackStatus } from "expo";
import { Item } from "../../reducers/itemsById";
import { AudioPlayerContext } from "./Context";

type SetPlaylist = (name: string, items: Item[], index: number) => void;
type SetSong = (uri: string, name: string) => Promise<void>;

export interface IAudioPlayerState {
  currentIndex: number;
  currentSongName: string;
  currentSongURL: string;
  getSeekSliderPosition: () => number;
  getSongName: () => string;
  getSongURL: () => string;
  getTimeStamp: () => string;
  isBuffering: boolean;
  isPlaying: boolean;
  isSeeking: boolean;
  items: Item[];
  muted: boolean;
  name: string;
  nextSong: () => void;
  onMutePressed: () => void;
  onPlayPausePressed: () => void;
  onSeekSliderChange: () => void;
  onSeekSliderComplete: (value: number) => Promise<void>;
  onStopPressed: () => void;
  playbackInstanceDuration: number | undefined;
  playbackInstancePosition: number | undefined;
  playerType: "song" | "playlist";
  previousSong: () => void;
  seekSong: (index: number) => void;
  setPlaylist: SetPlaylist;
  setSong: SetSong;
}

export class AudioPlayerProvider extends React.Component<{}, IAudioPlayerState> {
  isSeeking = false;
  playbackInstance: PlaybackObject | null = null;
  constructor(props: any) {
    super(props);
    this.state = {
      currentIndex: 0,
      currentSongName: "",
      currentSongURL: "",
      getSeekSliderPosition: this.getSeekSliderPosition,
      getSongName: this.getSongName,
      getSongURL: this.getSongURL,
      getTimeStamp: this.getTimestamp,
      isBuffering: false,
      isPlaying: false,
      isSeeking: false,
      items: [],
      muted: false,
      name: "",
      nextSong: this.nextSong,
      onMutePressed: this.onMutePressed,
      onPlayPausePressed: this.onPlayPausePressed,
      onSeekSliderChange: this.onSeekSliderValueChange,
      onSeekSliderComplete: this.onSeekSliderSlidingComplete,
      onStopPressed: this.onStopPressed,
      playbackInstanceDuration: undefined,
      playbackInstancePosition: undefined,
      playerType: "song" as "song" | "playlist",
      previousSong: this.previousSong,
      seekSong: this.seekSong,
      setPlaylist: this.setPlaylist,
      setSong: this.setSong
    };
  }

  setSong: SetSong = async (uri, name) => {
    this.setState({
      currentSongName: name,
      currentSongURL: uri,
      playerType: "song"
    });
    await this.playSong(uri);
  };

  setPlaylist: SetPlaylist = (name, items, index = 0) => {
    this.setState({
      items,
      name,
      currentIndex: index,
      playerType: "playlist",
    });
    this.playSong(items[index].link);
  };

  playSong = async (uri: string): Promise<void> => {
    if (this.playbackInstance !== null) {
      await this.playbackInstance.unloadAsync();
      this.playbackInstance = null;
    }
    const { sound } = await (Audio.Sound as any).createAsync(
      { uri },
      { shouldPlay: true },
      this.onPlaybackStatusUpdate
    );

    this.playbackInstance = sound;
  };

  previousSong = (): void => {
    this.setState(({ currentIndex }: any) => {
      return {
        currentIndex: currentIndex === 0 ? currentIndex : currentIndex - 1
      };
    }, () => {
      this.playSong(this.getSongURL());
    });
  };

  seekSong = (index: number): void => {
    this.setState({ currentIndex: index });
    this.playSong(this.state.items[index].link);
  };

  nextSong = (): void => {
    this.setState(({ currentIndex, items }: any) => {
      const index = currentIndex === items.length - 1 ? 0 : currentIndex + 1;
      return {
        currentIndex: index
      };
    }, () => {
      this.playSong(this.getSongURL());
    });
  };

  onPlayPausePressed = (): void => {
    if (this.playbackInstance != null) {
      if (this.state.isPlaying) {
        this.playbackInstance.pauseAsync();
      } else {
        this.playbackInstance.playAsync();
      }
    }
  };

  // prevent flickering of seek bar and pause button
  shouldComponentUpdate(nextProps: Readonly<{}>, nextState: Readonly<IAudioPlayerState>, nextContext: any): boolean {
    return !this.isSeeking;
  }

  onPlaybackStatusUpdate = (status: PlaybackStatus): void => {
    if (status.isLoaded) {
      this.setState({
        ...this.state,
        isBuffering: status.isBuffering,
        isPlaying: status.isPlaying,
        muted: status.isMuted,
        playbackInstanceDuration: status.durationMillis,
        playbackInstancePosition: status.positionMillis,
      });
    }
  };

  onStopPressed = (): void => {
    if (this.playbackInstance !== null) {
      this.playbackInstance.stopAsync();
    }
  };

  onMutePressed = (): void => {
    if (this.playbackInstance !== null) {
      this.playbackInstance.setIsMutedAsync(!this.state.muted);
    }
  };

  onSeekSliderValueChange = (): void => {
    if (this.playbackInstance !== null && !this.isSeeking) {
      this.setState({ isPlaying: false }, () => this.isSeeking = true);

      this.playbackInstance && this.playbackInstance.pauseAsync();
    }
  };

  onSeekSliderSlidingComplete = async (value: number): Promise<void> => {
    if (this.playbackInstance !== null) {
      const { playbackInstanceDuration } = this.state;
      if (playbackInstanceDuration) {
        const seekPosition = value * playbackInstanceDuration;
        await this.playbackInstance.playFromPositionAsync(seekPosition);
        this.isSeeking = false;
      }
    }
  };

  getSeekSliderPosition = (): number => {
    const { playbackInstanceDuration, playbackInstancePosition } = this.state;
    if (this.playbackInstance && playbackInstanceDuration && playbackInstancePosition) {
      return playbackInstancePosition / playbackInstanceDuration;
    }
    return 0;
  };

  getSongName = (): string => {
    const { currentIndex, currentSongName, items, playerType } = this.state;
    if (playerType === "song") {
      return currentSongName;
    } else {
      return items[currentIndex].title;
    }
  };

  getSongURL = (): string => {
    const { currentIndex, currentSongURL, items, playerType } = this.state;
    if (playerType === "song") {
      return currentSongURL;
    } else {
      return items[currentIndex].link;
    }
  };

  getMMSSFromMillis = (millis: number): string => {
    const totalSeconds = millis / 1000;
    const seconds = Math.floor(totalSeconds % 60);
    const minutes = Math.floor(totalSeconds / 60);

    const padWithZero = (time: number) => {
      if (time < 10) {
        return `0${time}`;
      }
      return time.toString();
    };
    let timeFormat = `${padWithZero(minutes)}:${padWithZero(seconds)}`;
    if (minutes > 59) {
      const hours = Math.floor(minutes / 60);
      timeFormat = `${hours}:${timeFormat}`;
    }
    return timeFormat;
  };

  getTimestamp = (): string => {
    if (
      this.playbackInstance != null &&
      this.state.playbackInstancePosition != null &&
      this.state.playbackInstanceDuration != null
    ) {
      return `${this.getMMSSFromMillis(
        this.state.playbackInstancePosition
      )} / ${this.getMMSSFromMillis(this.state.playbackInstanceDuration)}`;
    }
    return "00:00 / 00:00";
  };

  render() {
    return (
      <AudioPlayerContext.Provider value={this.state}>
        {this.props.children}
      </AudioPlayerContext.Provider>
    );
  }
}

export default AudioPlayerProvider;
