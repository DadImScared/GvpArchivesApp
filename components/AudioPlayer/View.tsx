import * as React from "react";
import { View as Wrapper } from "react-native";

import Audio from "./Audio";
import ButtonGroup from "./ButtonGroup";
import DurationSeeker from "./DurationSeeker";
import Title from "./Title";

const View = () => (
  <Wrapper>
    <Audio />
    <Title/>
    <DurationSeeker/>
    <ButtonGroup/>
  </Wrapper>
);

export default View;
