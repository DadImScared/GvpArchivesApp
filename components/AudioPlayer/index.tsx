import * as React from "react";
import { StyleSheet, View } from "react-native";

import { withAudioPlayer } from "../withAudioPlayer";
import ButtonGroup from "./ButtonGroup";
import DurationSeeker from "./DurationSeeker";
import { IAudioPlayerState } from "./Provider";
import Title from "./Title";

interface IProps {
  audio: IAudioPlayerState;
}

const styles = StyleSheet.create({
  wrapper: {
    backgroundColor: "white",
    bottom: 0,
    padding: 8,
    position: "absolute",
    width: "100%",
  }
});

export class AudioPlayer extends React.Component<IProps> {
  render() {
    const {
      audio: {
        isSeeking,
        isPlaying,
        onPlayPausePressed,
        getSongName,
        getSeekSliderPosition,
        getTimeStamp,
        onSeekSliderChange,
        onSeekSliderComplete

      }
    } = this.props;
    return (
      <View style={styles.wrapper}>
        <Title getSongName={getSongName} />
        <DurationSeeker
          isSeeking={isSeeking}
          getSeekSliderPosition={getSeekSliderPosition}
          getTimeStamp={getTimeStamp}
          onSeekSliderChange={onSeekSliderChange}
          onSeekSliderComplete={onSeekSliderComplete}
        />
        <ButtonGroup
          onPlayPausePressed={onPlayPausePressed}
          isPlaying={isPlaying}
        />
      </View>
    );
  }
}

export default withAudioPlayer(AudioPlayer) as any; // property error without "any"
