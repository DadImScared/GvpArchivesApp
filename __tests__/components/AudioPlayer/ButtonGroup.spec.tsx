import * as React from "react";
import * as renderer from "react-test-renderer";

import { ButtonGroup } from "../../../components/AudioPlayer/ButtonGroup";

describe("ButtonGroup", () => {
  it("should render", () => {
    const tree = renderer.create(<ButtonGroup playing={true} togglePlaying={jest.fn()} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
