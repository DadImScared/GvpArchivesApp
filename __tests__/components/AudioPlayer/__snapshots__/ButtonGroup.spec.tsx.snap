// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup should render 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "flex-start",
        "backgroundColor": "",
        "borderBottomWidth": null,
        "borderColor": "#007aff",
        "borderLeftWidth": null,
        "borderRadius": 5,
        "borderRightWidth": null,
        "borderTopWidth": null,
        "borderWidth": undefined,
        "elevation": 0,
        "flexDirection": "row",
        "height": 45,
        "justifyContent": "space-between",
        "opacity": 1,
        "paddingBottom": 6,
        "paddingTop": 6,
        "shadowColor": null,
        "shadowOffset": null,
        "shadowOpacity": null,
        "shadowRadius": null,
      }
    }
  >
    <Icon
      name="pause"
      style={
        Object {
          "color": "#3F51B5",
        }
      }
    />
  </View>
</View>
`;
